### ISSUE TITLE
Inconsistent error on return value type inferencing

### ISSUE THREAD

**Issue by haarisr** [microsoft/pyright#8667]
**Describe the bug**
When returning a dictionary, with keys as subclass type of a parent type, there is different behavious when
1. The dict is returned and instatiated at the same time 
2. The dict is assigned to a variable and the variable is returned 

**Code or Screenshots**
Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.374&code=MQBwngTglg5gFgFwFwAIDODoGMEFgBQWANgIZpooAiJCJSBKjKIZaBBxrKAkgLYkwApgApqtAJT18TZq3aFS5FAFkyAa1E0Skhkxbl5AE0EAzFIahoWCLHGEFxKALQA%2Bc1BwBtBGBCDPYiQAugA0GBBBUjIQgggArhAAdigA3royjHwCgqgARFD8Qrkh6RmqaGp5-BW5pQC%2BRqbuVjS2APqCEBAA9hD2%2BI6u7l4%2BfgFaoeGRpTEUALyppTJZQnkF2cVLTOWVKLnVarXSTA3HjDHxSSizQA)

```python
#pyright: strict
class Data:
    pass

class Image(Data):
    pass

class Mask(Data):
    pass

def dispatch(
) -> dict[type[Data],str]:
    return {
        Image: "image",
        Mask: "mask"
    }

def dispatch_error(
) -> dict[type[Data],str]:
    res = {
        Image: "image",
        Mask: "mask"
    }
    return res
```

**VS Code extension or command-line**
Latest pyright on the command line

**Comment by erictraut** [comment]
Pyright's behavior here is correct.

In your code sample, the dict expression in `dispatch_error` function uses default [dictionary inference rules](https://microsoft.github.io/pyright/#/type-inference?id=dictionary-expressions). In the `dispatch` function, default inference rules are overridden by [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). If you want the `dispatch_error` function to type check without errors, you can add an explicit type declaration for the `res` variable.

```python
def dispatch_error() -> dict[type[Data], str]:
    res: dict[type[Data], str] = {Image: "image", Mask: "mask"}
    return res
```
